Fix out step not responding with version data

crusstu opened the pull request:
I took a stab at fixing #44, but would definitely appreciate feedback.

Design-wise I wasn't sure if the logic to pull out the version from the blob name should be turned into its own function and if so where it should live. The implementation in `check` loops over all the blobs in the selected blobstore and does some other interleaved work so I'm not sure if the two callsites can have the common logic extracted.

Closes #44.

crusstu committed:
Fix out step not responding with version data

For blobs matched by regexes, the check step pulls out the version
number from the blob name and adds it to the response. This commit adds
the same behaviour to the out step so that Concourse sees versions for
the same blob produced by the out and check steps as being equal by
having both `path` and `version` fields with the same values.

Closes #44.

crusstu commented on cmd/out/main.go:
```diff
@@ -49,6 +53,27 @@ func main() {
+		}
+
+		matches := matcher.FindStringSubmatch(blobName)
+		if len(matches) >= 2 {
```
I wasn't sure if `len(matches) < 2` if the regex doesn't match anything in the blob name should be a fatal error, or if this implementation was sufficient.

christianang replied:
I think not failing here is okay. If it causes unforeseen issues to not explicitly error we can add it in later, but I don't expect this to happen. It also keeps old behavior (i.e didn't use to error if the regex didn't find a match), so in that regard it may also be valuable.

christianang commented on cmd/out/main.go:
```diff
@@ -49,6 +53,27 @@ func main() {
+			log.Fatal("failed to compile source configuration regex: ", err)
+		}
+
+		matches := matcher.FindStringSubmatch(blobName)
```
The `blobName` pulls from the params and isn't the actual path, which has the version in it. I think this code needs to move lower down to after the `out.UploadFileToBlobstore` call and matches on the `path`.

Noticed this by running the `./acceptance` tests. Specifically [this test](https://github.com/pivotal-cf/azure-blobstore-resource/blob/master/acceptance/out_test.go#L149) if you'd like one to focus. Unfortunately don't have GitHub PR actions to run the tests so you'll have to run this locally. It should be fairly straight forward though, you will need a storage account name and the storage account key the tests can use to upload/download files to/from Azure. More info [here](https://github.com/pivotal-cf/azure-blobstore-resource/blob/master/acceptance/README.md). Happy to answer any questions if you get stuck.

crusstu replied:
Thanks for the feedback, Christian! I'd taken a look at the unit tests but completely missed the `./acceptance` directory, I'll have a look at enhancing the test cases to make sure we check for the version on the output and get this fixed up.

christianang commented:
Thanks for putting the work into this PR! I tried this locally and ran into an issue, once the above is fixed and I've tested that the change works I think it can be merged.

re: design question. I think your design decisions made sense. I'm fine with the duplication and if it occurs in more places after this we can reconsider if it should be abstracted.

Feel free to ask questions here if you have any trouble. Thanks!

crusstu commented:
Hey @christianang, thanks again for the feedback; I think I've addressed the points you raised.

I see what you mean about `blobName`, so I moved the version extraction logic to operate on the `path` returned by the blob upload instead.

I also added a test for the version number and was able to validate it with the `./acceptance` tests, so we should be good on the tests front.

Let me know if there's anything I've missed!

christianang commented:
Looks great! Merged, once it makes it through CI I'll cut a new release.

christianang commented:
New release has been cut: https://github.com/pivotal-cf/azure-blobstore-resource/releases/tag/v0.13.0
